---
title: "CMap Drug Safety Challenge"
author: "Joaquim Aguirre-Plans"
date: "March 28, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
### Install packages ###
# Install Bioconductor to install "rhdf5", a dependency of cmapR
#install.packages("BiocManager")
#BiocManager::install("rhdf5", version = "3.8")
# cmapR has been installed manually by downloading the package on github (https://github.com/cmap/cmapR) and running the following commands:
# R CMD build cmapR
# R CMD check cmapR_1.0.1.tar.gz
#install.packages("/home/quim/R/x86_64-pc-linux-gnu-library/3.5/cmapR_1.0.1.tar.gz", type="source", repos=NULL)
# Install GGPlot2 and its dependencies
#install.packages('Rcpp', dependencies = TRUE)
#install.packages('ggplot2', dependencies = TRUE)
#install.packages('data.table', dependencies = TRUE)
# Install ggbiplot to plot the PCA
#library(devtools)
#install_github("vqv/ggbiplot")

### Load packages ###
library(cmapR)
library(ggplot2)
library(ggbiplot)
```

```{r, include=FALSE}
### Define files ###
#setwd("/sbi/users/interchange/emre/quim/camda")
#dilirank_file2 <- "/sbi/users/interchange/emre/quim/camda/l1000_1314pert_inames.872matchedBROAD_ids-dilirank.rda" #File of compounds info matching cell painting image data
drugs_file <- "/home/quim/PHD/Projects/camda/camda_data/CAMDA_l1000_1314compounds-pert_iname.rda"
drug_info_file <- "/sbi/users/interchange/emre/quim/camda/CAMDA_l1000_1314compounds-info.rda"
dilirank_file <- "/home/quim/PHD/Projects/camda/camda_data/CAMDA_l1000_1314compounds-dilirank.v2.rda"
expression_file <- "/sbi/users/interchange/emre/quim/camda/CAMDA_l1000_1314compounds-GSE92742_Level5_gct.rda"
```


## Explore files

First I load the files and explore the objects loaded

```{r}
load(drugs_file)
load(drug_info_file)
load(dilirank_file)
load(expression_file) # Requires cmapR
```

### selDrugs

selDrugs contains the drugs that we work with:

```{r}
summary(selDrugs)
```

### selDrugsInfo13

selDrugsInfo13 contains the information about the InChiKey and SMILES of the drugs:

```{r}
summary(selDrugsInfo13)
```

### drank.sel

drank.sel contains the information about the DILIRank and additional information about the drugs:

```{r}
summary(drank.sel)
```


In the drank.sel object we have extensive information about each drug:

* **pert_iname**: "perturbation name", the name of the drug
```{r}
drank.sel$pert_iname[0:10]
```

* **Other identifiers of the drug**:
```{r}
drank.sel$Compound.Name[0:10]
drank.sel$LTKBID[0:10]
drank.sel$PubChem_CID[0:10]
```

* **The SMILES**:
```{r}
drank.sel$SMILES[0:10]
```

* **DILIConcern**: The classes of concert in DILIRank
```{r}
drank.sel$DILIConcern[drank.sel$DILIConcern == "No-DILI-concern"] <- "No-DILI-Concern" # Correct the 4 entries with lowercase concern
# Assign simple names to DILIConcern labels
drank.sel$DILIConcern[drank.sel$DILIConcern == "Most-DILI-Concern"] <- "Most"
drank.sel$DILIConcern[drank.sel$DILIConcern == "Less-DILI-Concern"] <- "Less"
drank.sel$DILIConcern[drank.sel$DILIConcern == "No-DILI-Concern"] <- "No"
# Check number of drugs in each class
table(drank.sel$DILIConcern)
```

* **Additional information related with the DILIRank concern**:
```{r}
table(drank.sel$Severity.Class) # 7-8 = Most, 2-6 = Less, 0 = No
table(drank.sel$vDILIConcern)
table(drank.sel$AdjudicatedDILI)
table(drank.sel$Label.Section)
```

* **Other measures of DILI concern**:
```{r}
table(drank.sel$Greene_Annotation)
table(drank.sel$Sakatis_Annotation)
table(drank.sel$Xu_Annotation)
table(drank.sel$Zhu_Annotation)
```


### gct

gct is a cmapR object. In the gct object, the rows are genes (in Entrez Gene ID) and the columns are samples of cell lines after drug administration in different conditions:

```{r}
summary(gct)
summary(gct@rid)
summary(gct@cid)
```

In the gct object there is the "cdesc" field which provides us with different types of information of the samples:

```{r}
summary(gct@cdesc)
```

Here I provide a small description of the most relevant fields:

* **pert_id**: "perturbation ID", the identifier of the drug
```{r}
gct@cdesc$pert_id[0:10]
```
* **pert_iname**: "perturbation name", the name of the drug
```{r}
gct@cdesc$pert_iname[0:10]
```
* **pert_type**: "perturbation type"
```{r}
unique(gct@cdesc$pert_type)
```
* **cell_id**: "cell identifier", type of cell line
```{r}
unique(gct@cdesc$cell_id)
```
* **pert_dose**: "perturbation dose", dose of the drug administration
```{r}
unique(gct@cdesc$pert_dose)[0:20]
```
* **pert_dose_unit**: "perturbation dose unit", unit of the dose
```{r}
unique(gct@cdesc$pert_dose_unit)
```
* **pert_idose**: "perturbation idose", dose + unit
```{r}
unique(gct@cdesc$pert_idose)
```
* **pert_time**: "perturbation time", duration of the dose administration
```{r}
unique(gct@cdesc$pert_time)
```
* **pert_time_unit**: "perturbation time unit", unit of time
```{r}
unique(gct@cdesc$pert_time_unit)
```
* **pert_itime**: "perturbation itime", time + unit
```{r}
unique(gct@cdesc$pert_itime)
```
* **distil_id**: ID of an individual replicate profile
```{r}
gct@cdesc$distil_id[0:10]
```


## Subset the gene expression data

There are three relevant parameters if we want to subset the gene expression data of a given drug:

* cell_id: The type of cell line
* pert_idose: The quantity of drug dose
* pert_itime: The duration of drug dose

Let's analyse the example of the drug voriconazole that is actually in the DILI category "Most-DILI-Concern":

```{r}
drank.sel$DILIConcern[drank.sel$pert_iname=="voriconazole"]
```

We subset the drug information using the following command:

```{r}
idx <- which(gct@cdesc$pert_iname=="voriconazole")
x <- subset.gct(gct, cid=idx)
```

And now we can analyse the distribution of expression values (z-score) in each sample for the gene 1 for each time and cell line:

```{r}
labs = apply(x@cdesc[,c("cell_id","pert_itime")], 1, function(y) { paste0(y, collapse="-")})
s = split(x@mat[1,], labs)
boxplot(s,horizontal = T)
```


We observe that there is a great variation in expression for one gene, so to unify all the samples in one value will be a challenge!


## Analysis of gene expression similarity between pairs drugs in different DILI classes

We will unify the samples of each drug in every gene by two methods:
* Considering differentially expressed the genes with the absolute value of the mean of all the samples higher than 1.
* Considering differentially expressed the genes with more than 50% of the samples having an absolute z-score higher than 1.

### Considering the mean of all the samples for each gene

```{r}
fraction_samples <- 0.33
# Create new dataframe to store the mean expression of the samples of a drug in every gene
means_df <- data.frame(matrix(ncol = length(drank.sel$pert_iname), nrow = length(gct@rid)))
colnames(means_df) <- drank.sel$pert_iname
rownames(means_df) <- gct@rid

# Create new dataframe to store 1 if >50% of the samples of a drug in a gene are abs(z-score)>, 0 otherwise
diff_df <- data.frame(matrix(ncol = length(drank.sel$pert_iname), nrow = length(gct@rid)))
colnames(means_df) <- drank.sel$pert_iname
rownames(means_df) <- gct@rid

# Iterate over each drug
for (drug in drank.sel$pert_iname){
  # Get the subset of samples of the drug
  idx <- which(gct@cdesc$pert_iname==drug)
  x <- subset.gct(gct, cid=idx)
  
  # For each gene, calculate the mean z-score among all the samples of the drug
  means<- rowMeans(x@mat)
  means_df[,drug] <- means
  
  # Also check for each gene if >=50% of the samples of the drug have an abs(z-score) above 1. If so, return 1, otherwise, return 0
  diff <- apply(x@mat, 1, function(y){
      acc <- sum(abs(y)>1)
      if (acc>=(ncol(x@mat)*fraction_samples)){
        return(1)
      }else{
        return(0)
      }
    })
  diff_df[,drug] <- diff
}

# Function to calculate the Jaccard Coefficient using as input values lists of 0s or 1s
Jaccard = function (x, y) {
    M.11 = sum(x == 1 & y == 1)
    M.10 = sum(x == 1 & y == 0)
    M.01 = sum(x == 0 & y == 1)
    if((M.11 + M.10 + M.01)==0){
      return (0)
    }else{
      return (M.11 / (M.11 + M.10 + M.01))
    }
}

# Create a dataframe to store the jaccard coefficient of each pair of drugs (using mean)
mean_jac_df <- data.frame(matrix(ncol = 6, nrow = 0))
colnames(mean_jac_df) <- c('drug1', 'group1', 'drug2', 'group2', 'grouppair', 'jaccard')

# Create a dataframe to store the jaccard coefficient of each pair of drugs (using >50% of genes with abs(z-score)>1)
diff_jac_df <- data.frame(matrix(ncol = 6, nrow = 0))
colnames(diff_jac_df) <- c('drug1', 'group1', 'drug2', 'group2', 'grouppair', 'jaccard')

# For each pair of drugs, calculate the jaccard coefficient using the data of both methods (mean or >50%)
drugs <- drank.sel$pert_iname
for (i in 1:length(drugs)){
  drug1 <- drugs[i]
  group1 <- drank.sel$DILIConcern[drank.sel$pert_iname==drug1]
  for (j in i:length(drugs)){
    drug2 <- drugs[j]
    group2 <- drank.sel$DILIConcern[drank.sel$pert_iname==drug2]
    grouppair <- paste(sort(c(group1, group2)),collapse="-")
    if (drug1 == drug2){
      next
    }
    
    # For abs(mean)>1
    means1 <- means_df[,drug1]
    means2 <- means_df[,drug2]
    means1[abs(means1)<1] <- 0
    means1[abs(means1)>1] <- 1
    means2[abs(means2)<1] <- 0
    means2[abs(means2)>1] <- 1
    jacindex_mean <- Jaccard(means1, means2)
    if ( is.na(jacindex_mean) | jacindex_mean > 1 | jacindex_mean < 0 ){
      print(paste(drug1, drug2, jacindex_mean))
    }
    mean_jac_df[nrow(mean_jac_df)+1,] <- c(drug1, group1, drug2, group2, grouppair, jacindex_mean)
    
    # For >50% abs(z-score)>1
    jacindex_diff <- Jaccard(diff_df[,drug1], diff_df[,drug2])
    if ( is.na(jacindex_diff) | jacindex_diff > 1 | jacindex_diff < 0 ){
      print(paste(drug1, drug2, jacindex_diff))
    }
    diff_jac_df[nrow(diff_jac_df)+1,] <- c(drug1, group1, drug2, group2, grouppair, jacindex_diff)
    
    #print(paste(drug1, drug2))
  }
}

# Convert the jaccard to a numeric variable
mean_jac_df$grouppair <- as.character(mean_jac_df$grouppair)
mean_jac_df$jaccard <- as.numeric(mean_jac_df$jaccard)
diff_jac_df$grouppair <- as.character(diff_jac_df$grouppair)
diff_jac_df$jaccard <- as.numeric(diff_jac_df$jaccard)
```

Here we show the boxplot of the jaccard coefficient between pairs of drugs depending on the number of differentially expressed genes that they have (using the *mean* method).
```{r}
# Create a boxplot showing the result of the jaccard coefficient for the pairs of drugs belonging to certain pairs of groups
p <- ggplot(mean_jac_df, aes(x=grouppair, y=jaccard, fill=grouppair)) +
  geom_boxplot()
#p <- ggplot(mean_jac_df, aes(factor(grouppair), jaccard)) + 
#  geom_violin(aes(fill = grouppair))
p
```


Here we show the boxplot of the jaccard coefficient between pairs of drugs depending on the number of differentially expressed genes that they have (using the *>50% abs(z-score)>1* method).

```{r}
p <- ggplot(diff_jac_df, aes(x=grouppair, y=jaccard, fill=grouppair)) +
  geom_boxplot()
p
```

### Considering only one specific condition

We will try to analyse the similarity in the expression of the genes comparing the pairs of drugs of the different DILIrank groups. But this time, we will choose specific conditions:

* **Cell ID**: HEPG2
* **Dose**: 10 µM
* **Time**: 6 h

```{r}
# Conditions
cell_id <- "HEPG2"
dose <- "10 µM"
time <- "6 h"


# Create new dataframe to store 1 if abs(z-score)>1, 0 otherwise
spec_df <- data.frame(matrix(ncol = length(drank.sel$pert_iname), nrow = length(gct@rid)))
colnames(spec_df) <- drank.sel$pert_iname
rownames(spec_df) <- gct@rid

# Create new dataframe to store gene expression values for pca
pca_df <- data.frame(matrix(ncol = length(gct@rid), nrow = length(drank.sel$pert_iname)))
colnames(pca_df) <- gct@rid
rownames(pca_df) <- drank.sel$pert_iname

# Iterate over each drug
for (drug in drank.sel$pert_iname){
  # Get the subset of samples of the drug
  idx <- which(gct@cdesc$pert_iname==drug)
  x <- subset.gct(gct, cid=idx) # Subset by drug
  idx_c <- which(x@cdesc$cell_id==cell_id)
  x_c <- subset.gct(x, cid=idx_c) # Subset by cell ID
  idx_cd <- which(x_c@cdesc$pert_idose==dose)
  x_cd <- subset.gct(x_c, cid=idx_cd) # Subset by dose
  idx_cdt <- which(x_cd@cdesc$pert_itime==time)
  x_cdt <- subset.gct(x_cd, cid=idx_cdt) # Subset by time
  #print(paste(drug,length(x_cdt@cid)))
  if (length(x_cdt@cid) > 0){
    # Check for each gene if abs(z-score)>1. If so, return 1, otherwise, return 0
    diff <- apply(x_cdt@mat, 1, function(y){
        mn <- mean(y)
        ifelse(abs(mn)>1, 1, 0)
      })
    spec_df[,drug] <- diff
    pca_df[drug,] <- rowMeans(x_cdt@mat)
  } else {
    spec_df[drug] <- NULL
    pca_df[drug] <- NULL
  }
}


# Function to calculate the Jaccard Coefficient using as input values lists of 0s or 1s
Jaccard = function (x, y) {
    M.11 = sum(x == 1 & y == 1)
    M.10 = sum(x == 1 & y == 0)
    M.01 = sum(x == 0 & y == 1)
    if((M.11 + M.10 + M.01)==0){
      return (0)
    }else{
      return (M.11 / (M.11 + M.10 + M.01))
    }
}

# Create a dataframe to store the jaccard coefficient of each pair of drugs (using mean)
spec_jac_df <- data.frame(matrix(ncol = 6, nrow = 0))
colnames(spec_jac_df) <- c('drug1', 'group1', 'drug2', 'group2', 'grouppair', 'jaccard')

# For each pair of drugs, calculate the jaccard coefficient
drugs <- colnames(spec_df)
for (i in 1:length(drugs)){
  drug1 <- drugs[i]
  group1 <- drank.sel$DILIConcern[drank.sel$pert_iname==drug1]
  for (j in i:length(drugs)){
    drug2 <- drugs[j]
    group2 <- drank.sel$DILIConcern[drank.sel$pert_iname==drug2]
    grouppair <- paste(sort(c(group1, group2)),collapse="-")
    if (drug1 == drug2){
      next
    }
    
    jacindex <- Jaccard(spec_df[,drug1], spec_df[,drug2])
    if ( is.na(jacindex) | jacindex > 1 | jacindex < 0 ){
      print(paste(drug1, drug2, jacindex))
    }
    spec_jac_df[nrow(spec_jac_df)+1,] <- c(drug1, group1, drug2, group2, grouppair, jacindex)
    
  }
}

# Convert the jaccard to a numeric variable
spec_jac_df$grouppair <- as.character(spec_jac_df$grouppair)
spec_jac_df$jaccard <- as.numeric(spec_jac_df$jaccard)

# Plot the boxplot
p <- ggplot(spec_jac_df, aes(x=grouppair, y=jaccard, fill=grouppair)) +
  geom_boxplot()
p
```

Let's calculate the PCA:

```{r}
# Example with mtcars
#mtcars.pca <- prcomp(mtcars[,c(1:7,10,11)], center = TRUE,scale. = TRUE)
#summary(mtcars.pca)
#ggbiplot(mtcars.pca)
#mtcars.pca <- prcomp(mtcars[,c(1:7,10,11)], center = TRUE,scale. = TRUE)
#mtcars.country <- c(rep("Japan", 3), rep("US",4), rep("Europe", 7),rep("US",3), "Europe", rep("Japan", 3), rep("US",4), rep("Europe", 3), "US", rep("Europe", 3))
#ggbiplot(mtcars.pca,ellipse=TRUE, groups=mtcars.country)

# Define the groups of the drugs
groups <- c()
drugs <- rownames(pca_df)
for (i in 1:length(drugs)){
  drug <- drugs[i]
  groups[i] <- drank.sel$DILIConcern[drank.sel$pert_iname==drug]
}
# Calculate the PCA
pca_df <- pca_df[rowSums(is.na(pca_df)) != ncol(pca_df), ]
cmapdata.pca <- prcomp(pca_df, center = TRUE,scale. = TRUE)
summary(cmapdata.pca)
# Plot the PCA
ggbiplot(cmapdata.pca,ellipse=TRUE,var.axes=FALSE, groups=groups)
```


```{r, include=FALSE}
# in apply(), the second argument is the dimension: 2 means to apply the function to the dimension of the columns, while 1 applies it to the rows
#diff1 <- apply(means1, 2, function(x){sum(abs(x)>1)})
#diff2 <- apply(means2, 2, function(x){sum(abs(x)>1)}) 
```
